test/utils.py

```diff
@@ -0,0 +1,310 @@
+PORT_BASE = 10000
+
+
+def get_docker_host_ip():
+    sysname = os.uname().sysname
+    if sysname != 'Darwin' and sysname != 'Linux':
+        raise ValueError(
+            "Failed to get ip in platforms other than Linux and macOS: {}".format(sysname)
+        )
+    cmd = 'ifconfig | grep -E "([0-9]{1,3}\\.){3}[0-9]{1,3}" | grep -v 127.0.0.1 | awk \'{ print $2 }\' | cut -f2 -d: | head -n1'
+    res = subprocess.run(cmd, shell=True, check=True, stdout=subprocess.PIPE, encoding='utf-8')
+    return res.stdout.rstrip()
+
+
+class CLIFailure(Exception):
+    pass
+
+
+class Node:
+
+    ip = None
+    port = None
+    rpc_port = None
+    peer_id = None
+    seed = None
+
+    def __init__(self, ip, port, rpc_port, seed):
+        self.ip = ip
+        self.port = port
+        self.rpc_port = rpc_port
+        self.seed = seed
+
+    def __repr__(self):
+        return "<Node seed={} peer_id={}>".format(
+            self.seed,
+            None if self.peer_id is None else self.peer_id[2:8],
+        )
+
+    @property
+    def name(self):
+        return f"whiteblock-node{self.seed}"
+
+    @property
+    def multiaddr(self):
+        return f"/ip4/{self.ip}/tcp/{self.port}/ipfs/{self.peer_id}"
+
+    def close(self):
+        subprocess.run(f"docker kill {self.name}", shell=True, stdout=subprocess.PIPE)
+        subprocess.run(f"docker rm -f {self.name}", shell=True, stdout=subprocess.PIPE)
+
+    def run(self, bootnodes=None):
+        """`bootnodes` should be a list of string. Each string should be a multiaddr.
+        """
+        self.close()
+        bootnodes_cmd = ""
+        if bootnodes is not None:
+            bootnodes_cmd = "-bootstrap -bootnodes={}".format(
+            ",".join(bootnodes),
+        )
+        cmd = "docker run -d --name {} -p {}:10000 -p {}:13000 ethresearch/sharding-p2p:dev sh -c \"./sharding-p2p-poc -loglevel=DEBUG -ip=0.0.0.0 -seed={} {}\"".format(
+            self.name,
+            self.port,
+            self.rpc_port,
+            self.seed,
+            bootnodes_cmd,
+        )
+        subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, check=True)
+
+    def cli(self, cmd, **kwargs):
+        cmd_list = cmd.split(' ')
+        cmd_quoted_param_list = ["'{}'".format(i) for i in cmd_list]
+        cmd_quoted_param_str = " ".join(cmd_quoted_param_list)
+        return subprocess.run(
+            [
+                "docker",
+                "exec",
+                "-t",
+                self.name,
+                "sh",
+                "-c",
+                "./sharding-p2p-poc '-client' {}".format(cmd_quoted_param_str),
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            encoding='utf-8',
+            **kwargs,
+        )
+
+    def cli_safe(self, cmd):
+        res = self.cli(cmd)
+        if res.returncode != 0:
+            print(res)
+            raise CLIFailure(
+                "exit_code={}, stdout={!r}, stderr={!r}".format(
```

Comment:
Question: What can `!r` do?

Reply:
I think it is similar to `{}`.format(repr(param))